src/osp/id_registry.h

```diff
@@ -0,0 +1,341 @@
+void IdRegistry<ID_T, NO_AUTO_RESIZE>::create(IT_T out, size_t count)
+{
+    if (m_deleted.count() < count)
+    {
+        // auto-resize
+        if constexpr (NO_AUTO_RESIZE)
+        {
+            throw std::runtime_error(
+                    "Reached max capacity with automatic resizing disabled");
+        }
+        else
+        {
+            reserve(std::max(capacity() + count, capacity() * 2));
+        }
+    }
+
+    m_deleted.take<IT_T, ID_T>(out, count);
+}
+
+template<typename ID_T, bool NO_AUTO_RESIZE>
+template <typename FUNC_T>
+void IdRegistry<ID_T, NO_AUTO_RESIZE>::for_each(FUNC_T func)
+{
+    // TODO: this is kind of inefficient and doesn't utilize the hierarchy
+    for (size_t i = 0; i < size(); i ++)
+    {
+        if ( ! m_deleted.test(i) )
+        {
+            func(ID_T(i));
+        }
+    }
+}
+
+//-----------------------------------------------------------------------------
+
+/**
+ * @brief Long term reference counted storage for IDs
+ */
+template<typename ID_T, typename REG_T>
+class IdStorage
+{
+    friend REG_T;
+
+public:
+    IdStorage() : m_id{ id_null<ID_T>() } { }
+    ~IdStorage() { assert( ! has_value() ); }
+
+    // Delete copy
+    IdStorage(IdStorage const& copy) = delete;
+    IdStorage& operator=(IdStorage const& copy) = delete;
+
+    // Allow move only if there's no ID stored
+    IdStorage(IdStorage&& move)
+    {
+        assert( ! has_value() );
+        m_id = std::exchange(move.m_id, id_null<ID_T>());
+    }
+    IdStorage& operator=(IdStorage&& move)
+    {
+        assert( ! has_value() );
+        m_id = std::exchange(move.m_id, id_null<ID_T>());
+        return *this;
+    }
+
+    operator ID_T() const noexcept { return m_id; }
+    constexpr ID_T value() const noexcept { return m_id; }
+    constexpr bool has_value() const noexcept { return m_id != id_null<ID_T>(); }
+
+private:
+    IdStorage(ID_T id) : m_id{ id } { }
+
+    ID_T m_id;
+
+}; // class IdStorage
+
+//-----------------------------------------------------------------------------
+
+/**
+ * @brief An IdRegistry that uses IdStorages to uniquely store Ids
+ */
+template<typename ID_T, bool NO_AUTO_RESIZE = false>
+class UniqueIdRegistry : IdRegistry<ID_T, NO_AUTO_RESIZE>
+{
+    using base_t = IdRegistry<ID_T, NO_AUTO_RESIZE>;
+public:
+
+    using Storage_t = IdStorage<ID_T, UniqueIdRegistry<ID_T, NO_AUTO_RESIZE> >;
+
+    using IdRegistry<ID_T, NO_AUTO_RESIZE>::IdRegistry;
+    using base_t::capacity;
+    using base_t::size;
+    using base_t::reserve;
+    using base_t::exists;
+    using base_t::for_each;
+
+    [[nodiscard]] Storage_t create()
+    {
+        return base_t::create();
+    }
+
+    void remove(Storage_t& rStorage)
+    {
+        base_t::remove(rStorage);
+        rStorage.release();
+    }
+}; // class UniqueIdRegistry
+
+
+//-----------------------------------------------------------------------------
+
+template<typename COUNT_T = uint8_t>
+class RefCount : std::vector<COUNT_T>
```

Comment:
Inherting from std:: containers is almost universally considered a bad-idea.

While I recognize that you're doing private inheritance, which is substantially less bad, consider instead doing aggregation.

Probably it's not the end of the world, so it's a judgement call, really.

Reply:
Hmm, would this work well as a mixin instead? ie. `RefCount< std::vector<uint8_t> >`